source-academy/cadet-frontend
# Added utility to report student infinite loops to Sentry

### Description

An infinite loop detector was added to js-slang (see https://github.com/source-academy/js-slang/pull/575).
In an effort to improve this detector, it will be helpful to collect code written by students with infinite loops as example/test data.

This utility will use Sentry to log infinite loop errors made by the student.
Both infinite loop errors detected by the infinite loop detector, and errors that slip past (and appear as stack overflows) will be sent to Sentry.

![image](https://user-images.githubusercontent.com/5654470/88294102-6ed71300-cd2e-11ea-80aa-9373cce7e669.png)

Note that all the infinite loops will be logged under a single Sentry issue (all events logged have the same Sentry 'fingerprint')

![image](https://user-images.githubusercontent.com/5654470/88294182-87472d80-cd2e-11ea-8180-cd9018532ed0.png)

Code can be found in Sentry's report (as above).

All of the data can be later downloaded using the [Sentry API](https://docs.sentry.io/api/events/get-group-events/).

### Type of change


### How to Test

1. Ensure REACT_APP_SENTRY_DSN is set to the correct url in the environment variables (.env file)
2. Run any program that will cause an infinite loop. Example: `function z(x){return z(x);}z(1);`
3. Check that the issue "Error: Infinite Loop" is added on the Sentry dashboard (similar to pic above)

### Checklist
